File: leetcode/string/1021.remove-outermost-parentheses.py
class Solution:
    def removeOuterParentheses(self, S: str) -> str:
        count = 0
        p = 0
        parts = []
        for i in range(len(S)):
            if S[i] == '(':
                count += 1
            elif S[i] == ')':
                count -= 1
            if count == 0:
                parts.append(S[p+1:i])
                p = i+1
                
        return ''.join(parts)

File: leetcodeJS/string/1021.remove-outermost-parentheses.js
/**
 * @param {string} S
 * @return {string}
 */
var removeOuterParentheses = function (S) {
  let balance = 0;
  let p = 0;
  let solution = [];
  for (let i = 0; i < S.length; i++) {
    if (S[i] == "(") balance++;
    else if (S[i] == ")") balance--;
    if (balance == 0) {
      solution.push(S.slice(p + 1, i));
      p = i + 1;
    }
  }
  return solution.join("");
};
